runtime/compiler/trj9/optimizer/ValueJProfiling.cpp

```diff
@@ -0,0 +1,883 @@
+   TR::Node *incCounter = TR::Node::create(address, counterType == TR::Int32 ? TR::iadd : TR::ladd, 2, load,
+      TR::Node::create(address, loadConst(counterType), 0, 1));
+   TR::Node *storeCounter = TR::Node::createWithSymRef(address, indirectStore(counterType), 2,
+      address, symRef);
+   storeCounter->setAndIncChild(1, incCounter);
+
+   return storeCounter;
+   }
+
+/*
+ * Generate a constant node matching the system's address width. Used for address calculations.
+ *
+ * \param example Example node to copy BCI.
+ * \param value Value to store in const.
+ */
+TR::Node *
+TR_ValueJProfiling::systemConst(TR::Node *example, uint64_t value)
+   {
+   TR::ILOpCodes constOp = TR::Compiler->target.is64Bit() ? TR::lconst : TR::iconst;
+   return TR::Node::create(example, constOp, 0, value);
+   }
+
+/*
+ * Its sometimes necessary to convert values to their rounded integer representation, in Int32
+ * or Int64, or to convert values to the system type, again in Int32 or Int64. This method
+ * handles those conversions as necessary. All conversions are unsigned.
+ *
+ * \param index The value to convert.
+ * \param dataType The datatype to convert to.
+ */
+TR::Node *
+TR_ValueJProfiling::convertType(TR::Node *index, TR::DataType dataType)
+   {
+   if (index->getDataType() == dataType)
+      return index;
+   else if (dataType == TR::Int64) 
+      {
+      switch (index->getDataType())
+         {
+         case TR::Int64:
+            return index;
+         case TR::Int32:
+            return TR::Node::create(index, TR::iu2l, 1, index);
+         case TR::Int16:
+            return TR::Node::create(index, TR::su2l, 1, index);
+         case TR::Int8:
+            return TR::Node::create(index, TR::bu2l, 1, index);
+         case TR::Address:
+            return TR::Node::create(index, TR::a2l, 1, index);
+         default:
+            TR_ASSERT_FATAL(0, "Unsupported source data type");
+         }
+      }
+   else if (dataType == TR::Int32)
+      {
+      switch (index->getDataType())
+         {
+         case TR::Int64:
+            return TR::Node::create(index, TR::l2i, 1, index);
+         case TR::Int32:
+            return index;
+         case TR::Int16:
+            return TR::Node::create(index, TR::su2i, 1, index);
+         case TR::Int8:
+            return TR::Node::create(index, TR::bu2i, 1, index);
+         case TR::Address:
+            return TR::Node::create(index, TR::a2i, 1, index);
+         default:
+            TR_ASSERT_FATAL(0, "Unsupported source data type");
+         }
+      }
+
+   TR_ASSERT_FATAL(0, "Unsupported target data type");
+   return NULL;
+   }
+
+TR::Node *
+TR_ValueJProfiling::computeHash(TR::Compilation *comp, TR_AbstractHashTableProfilerInfo *table, TR::Node *value, TR::Node *baseAddr)
+   {
+   if (!baseAddr)
+      baseAddr = TR::Node::aconst(value, (uintptr_t) table);
+
+   TR::ILOpCodes shiftOp = table->getDataType() == TR::Int32 ? TR::iushr : TR::lushr;
```

Comment:
might this look and read better as a setup of the whole thing for say Int64 and then a if Int32 overwrite them all and if something else assert? The repeated testing is a bit ugly and will be slower than necessary.